### DISCUSSION TITLE
Force HTTPS globally when using PHP attributes for routing

### DISCUSSION THREAD

**Discussion by gnito-org** [symfony/symfony#44652] [Q&A] [answered]
Prior to PHP 8 attributes, one could declare all your routes as HTTPS with the following in ``config/routes/annotations.yaml``:

```
controllers:
    resource: ../../src/Controller/
    type: annotation
    schemes: [ '%http_protocol%' ]
```
When [using PHP 8 attributes](https://symfony.com/doc/current/routing.html#creating-routes-as-attributes-or-annotations), ``config/routes/annotations.yaml`` does/should not exist, because, as per the docs, ``composer require doctrine/annotations`` is only needed if you want to use annotations.

The question is, if only using attributes, how does one globally force HTTPS for all your routes, anonymous and authenticated (including those routes generated by third-party bundles)?

**Comment by mbabker** [comment]
I deal with it at the webserver level myself (Apache or Nginx directives to redirect HTTP to HTTPS).

A new Symfony 6 install has a similar config file at `config/routes.yaml` based on https://github.com/symfony/recipes/blob/master/symfony/routing/6.0/config/routes.yaml, you could still add the `schemes` key there.

**Comment by gnito-org** [comment]
Yikes, [that](https://github.com/symfony/recipes/commit/c208077a18c4e36e18b105d4ebc9855181c02ab4) sounds like a bit of a hack in Symfony 6. The fact that there still is a ``type: annotation`` entry under the ``controllers`` key, when you only want to use attributes, can be highly confusing.

Nothing suggests that adding a ``schemes: [ '%http_protocol%' ]`` entry under the ``controllers`` key will force HTTPS when using only attributes for routes.

**Reply by mbabker** [reply]
In the Symfony components that support them, "annotation" covers both doc block based annotations and engine native attributes.  So, it's not really a hack, it's just the way things work right now.

https://github.com/symfony/symfony/pull/44615 might be of interest in this case (it adds "attribute" as its own type, but it doesn't make any functional code changes).

**Comment by gnito-org** [comment]
That will fix potential confusion if one can use ``attribute`` in the config files when you're actually using attributes. The PR is targeting 6.1, which means that fix will only be available in May 2022.

**Comment by gnito-org** [comment, marked as answer]
Just to wrap up this question.

Symfony 6 installs the following file by default, even if you do not issue the ``composer require doctrine/annotations`` command.

```
// config.routes.yaml
controllers:
    resource: ../src/Controller/
    type: annotation

kernel:
    resource: ../src/Kernel.php
    type: annotation
```
This config is required, even if you're only using attributes. Without this config, configuring routes using attributes does not work.
